Upgrade Surefire

apmoriarty opened the pull request:
Change Summary
- replaced `surefire-junit47` provider with `junit-jupiter-engine` and `junit-vintage-engine`
- `junit-jupiter-api`, `junit-jupiter-engine`, and `junit-vintage-engine` are all controlled by `version.junit-bom`
- surefire upgraded from `3.0.0-M6` to `3.2.5` to avoid a bug that couldn't parse junit versions with patch versions

apmoriarty committed:
Configure surefire to run JUnit5, Junit4 tests

apmoriarty committed:
Add missed module

ivakegg commented on warehouse/accumulo-extensions/pom.xml:
```diff
@@ -42,6 +42,7 @@
         <dependency>
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
+            <version>${version.junit}</version>
```
Do you need the version here?  Is it not inherited from the parent pom?

apmoriarty replied:
IntelliJ flagged this instance of junit as not matching the parent pom

jwomeara replied:
is junit:junit in the dependency management section of the parent pom?  if not, it should be, and you shouldn't need the version here.

ivakegg replied:
I wonder if the test scope somehow mucks with that.  You may need to add the test scope for that in the management section as well.  Just a guess.

jwomeara replied:
he already updated it to remove the version.  it wasn't needed.

i would actually advise against explicitly listing the test scope when managing dependencies.  if you specify a scope, it will only manage it for that scope, and OTHER imported dependency management sections could override your compile time version.

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/IngestTypePruningVisitor.java:
```diff
@@ -192,7 +192,7 @@ public Object visit(ASTAndNode node, Object data) {
         }
 
-        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+        for (int i = node.jjtGetNumChildren() - 1; i >= 0; i--) {
```
why go through the children backwards?

ivakegg replied:
Oh, because we are pruning children as we go I guess.  Perhaps a comment to that effect would be useful.

apmoriarty replied:
Definitely worth a comment

apmoriarty committed:
Update comments

jwomeara commented on pom.xml:
```diff
@@ -1146,6 +1146,18 @@
                 <version>${version.junit.jupiter}</version>
                 <scope>test</scope>
             </dependency>
+            <dependency>
```
Can we just update the pom to use the junit bom like we do in datawave-microservice-parent?  https://github.com/NationalSecurityAgency/datawave-parent/blob/main/pom.xml#L64

apmoriarty replied:
I'm leaving this as an explicit dependency.

apmoriarty replied:
I'll see about pulling in the `junit.bom`

jwomeara commented on warehouse/core/pom.xml:
```diff
@@ -198,6 +198,16 @@
             <artifactId>easymock</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
```
junit-jupiter-engine has a dependency on junit-jupiter-api.  so you probably only need junit-jupiter-engine here.

apmoriarty replied:
All the example docs I read had the explicit dependencies called out. I think that's better in the long run for readability.

ivakegg replied:
Not sure I agree with that.  That requires us to ensure we are using compatible versons.  It is simply more maintainance on our part.

apmoriarty replied:
I'll look at pulling that in

jwomeara reviewed (changes requested):
Looks good for the most part.  In addition to these changes, can you update the following tests to stop mixing junit4 and junit5 classes? (e.g. Assert vs Assertions, org.junit.Test vs org.junit.jupiter.api.Test, etc.)

- DatawaveJexlContextTest
- AncestorChildExpansionIteratorTest
- TFKeyUtilTest
- FiKeyUtilTest
- TermFrequencyIteratorTest
- SplitSelectorExtractorTest

apmoriarty committed:
Update unit tests to junit5

apmoriarty committed:
Add junit-bom depedency and use version of junit-bom for explicit jupiter and vintage dependencies

ivakegg commented on warehouse/pom.xml:
```diff
@@ -475,6 +475,18 @@
+                        <dependency>
+                            <groupId>org.junit.jupiter</groupId>
+                            <artifactId>junit-jupiter-engine</artifactId>
+                            <version>${version.junit.bom}</version>
```
isn't this already defined in the parent pom?

jwomeara replied:
Plugin dependency management isn't a thing, unfortunately.  If you want to avoid having to redefine this plugin in multiple poms you could just move the maven-surefire-plugin configuration from the warehouse and web-services poms to the pluginManagement entry for the datawave parent pom.  That way, all of the child modules will inherit the same surefire configuration, and you will only have to define it in one place.

Honestly, I'm not sure why we aren't already doing that.  The surefire plugin configuration for warehouse and web-services is identical.

ivakegg commented on pom.xml:
```diff
@@ -1468,9 +1486,14 @@
-                            <groupId>org.apache.maven.surefire</groupId>
-                            <artifactId>surefire-junit47</artifactId>
-                            <version>${version.surefire.plugin}</version>
+                            <groupId>org.junit.jupiter</groupId>
```
I am wondering if we add these dependencies in the parent pom to the dependency section as well as the dependency management sections, then do we need to add these dependencies to every pom?

apmoriarty replied:
I had tried that initially but the build fails locally saying that a version is required. Once that is resolved by adding the version here in the dependencies section the build fails later on with a failure to pull junit-bom from the packages repo. Something is amiss here because you should be able to declare `junit-bom` in the dependencies section with just the `groupId` and `artifactId`

jwomeara replied:
Plugin dependencies can't be managed, unfortunately.  https://issues.apache.org/jira/browse/MNG-2496

jwomeara commented on pom.xml:
```diff
@@ -1143,7 +1149,19 @@
+            <dependency>
+                <groupId>org.junit.vintage</groupId>
+                <artifactId>junit-vintage-engine</artifactId>
+                <version>${version.junit.bom}</version>
```
These should be removed.  The junit-bom is importing these.  Keeping them defeats the purpose of using the bom

jwomeara commented on core/pom.xml:
```diff
@@ -34,19 +29,16 @@
             <dependency>
                 <groupId>org.junit.jupiter</groupId>
                 <artifactId>junit-jupiter-api</artifactId>
-                <version>${version.junit.jupiter}</version>
                 <scope>test</scope>
             </dependency>
             <dependency>
                 <groupId>org.junit.jupiter</groupId>
                 <artifactId>junit-jupiter-engine</artifactId>
-                <version>${version.junit.jupiter}</version>
                 <scope>test</scope>
             </dependency>
             <dependency>
                 <groupId>org.junit.vintage</groupId>
                 <artifactId>junit-vintage-engine</artifactId>
-                <version>${version.junit.vintage}</version>
                 <scope>test</scope>
             </dependency>
```
You should remove these completely.  These dependencies are managed by the parent pom.  The child inherits the parent's dependency management, so if you had kept the versions in this would have been redundant.  Without the versions, I'm not sure what this accomplishes at all.  I think this is all that's in the dependency management section of this pom, so you can just remove dependency management completely.

jwomeara commented on core/pom.xml:
```diff
@@ -59,16 +51,19 @@
         <dependency>
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
+            <version>${version.junit}</version>
             <scope>test</scope>
         </dependency>
         <dependency>
             <groupId>org.junit.jupiter</groupId>
             <artifactId>junit-jupiter-engine</artifactId>
+            <version>${version.junit.bom}</version>
             <scope>test</scope>
         </dependency>
         <dependency>
             <groupId>org.junit.vintage</groupId>
             <artifactId>junit-vintage-engine</artifactId>
+            <version>${version.junit.bom}</version>
             <scope>test</scope>
         </dependency>
```
No versions needed here since these dependencies are managed by the parent.

apmoriarty committed:
Clean up extra jupiter dependency declarations

apmoriarty committed:
Additional module cleanup

apmoriarty committed:
Add back some dependencies that were removed..

apmoriarty committed:
add junit-jupiter-engine to common-test

add junit-jupiter-engine to warehouse-common

remove jupiter-api from warehouse pom

restore jupitier-api to ws-client pom

apmoriarty committed:
remove junit version from core pom.xml

jwomeara commented on warehouse/pom.xml:
```diff
@@ -444,6 +444,19 @@
             <groupId>gov.nsa.datawave</groupId>
             <artifactId>datawave-common-test</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.vintage</groupId>
+            <artifactId>junit-vintage-engine</artifactId>
+        </dependency>
```
Are these needed here?  and if so, shouldn't they be test scoped?

apmoriarty replied:
Yes these are needed and yes they should be test scoped.

jwomeara replied:
the warehouse module itself has no source code, which means it has no tests to compile, let alone run, so why do you think these are needed?  for what it's worth I removed all of the dependencies in this pom and everything built just fine...

jwomeara replied:
happy to approve once this is removed.

jwomeara commented on warehouse/query-core/src/test/java/datawave/audit/SplitSelectorExtractorTest.java:
```diff
@@ -1,23 +1,23 @@
 
 import datawave.webservice.query.QueryImpl;
 
-public class SplitSelectorExtractorTest {
```
curious why public was removed throughout.

apmoriarty replied:
SonarLint will flag public modifiers in JUnit5 classes and test methods. Using the default package visible is recommended to improve readability.

jwomeara commented on web-services/client/pom.xml:
```diff
@@ -171,11 +171,6 @@
             <artifactId>junit</artifactId>
             <scope>test</scope>
         </dependency>
-        <dependency>
-            <groupId>org.junit.jupiter</groupId>
-            <artifactId>junit-jupiter-api</artifactId>
-            <scope>test</scope>
-        </dependency>
```
why are we removing this?  if we're not going to add junit-jupiter-api to each module, why remove it from the ones that already include it?

apmoriarty replied:
I removed it because it was unused, whereas with the other modules there are tests that rely on the jupiter api. I can add back the unused dependency if you like.

jwomeara replied:
i guess what im saying is that this change wasn't required for the work you were doing.  if you're worried about scope creep, well i think this counts as that, right?  only change what you need

jwomeara commented on common-test/pom.xml:
```diff
@@ -63,5 +63,10 @@
         </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
```
junit-jupiter-engine

jwomeara commented on warehouse/common/pom.xml:
```diff
@@ -55,5 +55,10 @@
         </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
```
junit-jupiter-engine

jwomeara commented on warehouse/pom.xml:
```diff
@@ -444,6 +444,19 @@
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
```
junit-jupiter-engine

apmoriarty committed:
fix bad rebase

jwomeara reviewed (changes requested):
just need some stuff removed from the warehouse pom.  also, not sure why `public` was removed from all of the test methods

apmoriarty committed:
remove dependencies from warehouse pom

apmoriarty committed:
add back dependency on datawave-common-test

apmoriarty committed:
add junit-jupiter-engine to warehouse pom with test scope

apmoriarty committed:
Add jupiter-engine dependency to webservices pom and ssdeep-common pom

jwomeara commented on warehouse/pom.xml:
```diff
@@ -445,6 +445,11 @@
             <groupId>gov.nsa.datawave</groupId>
             <artifactId>datawave-common-test</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
```
You don't need this.

apmoriarty replied:
If you look at the recent commit history you can see how this was removed, and then added back in when other modules failed to build.

jwomeara replied:
i was wrong.  if we keep this, all child modules will inherig junit-jupiter-engine, which is what we want.  sorry for the confusion

apmoriarty replied:
all good!

jwomeara commented on warehouse/ssdeep-common/pom.xml:
```diff
@@ -24,5 +24,10 @@
             <artifactId>junit</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
```
does ssdeep have junit jupiter tests?

jwomeara replied:
i think this can go away

jwomeara replied:
wrong again

jwomeara commented on web-services/pom.xml:
```diff
@@ -506,6 +506,11 @@
             <groupId>gov.nsa.datawave</groupId>
             <artifactId>datawave-common-test</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
+        </dependency>
```
does web-services have junit jupiter tests?

jwomeara replied:
i was wrong.  let's keep this

apmoriarty committed:
remove duplicate junit-jupiter-engine dependency from warehouse-common pom, add jupiter-engine and vintage-engine to failsafe plugins in webservices and warehouse pom

jwomeara reviewed (approved):
ship it

apmoriarty committed:
Merge branch 'integration' into task/upgrade-surefire

drewfarris commented:
Would it be possible to squash commits like this in the future? The 17 commits made rebasing some feature branches really challenging when there were conflicts.

jwomeara commented:
I agree.  PR commits should be squashed when merging.

apmoriarty commented:
100% agree, when merging this one I forgot to change the default merge strategy. Perhaps that should be updated?

jwomeara commented:
You can't set a default merge strategy for a project. GitHub defaults to the last merge strategy you selected on a per user basis. So you need to manually select squash and merge the next time you merge a PR to reset your default.
